crates/engine/tree/src/tree/mod.rs

```diff
@@ -166,29 +168,104 @@ impl InMemoryStateImpl {
         numbers: HashMap<u64, B256>,
         pending: Option<State>,
     ) -> Self {
-        Self { blocks, numbers, pending }
+        Self {
+            blocks: RwLock::new(blocks),
+            numbers: RwLock::new(numbers),
+            pending: RwLock::new(pending),
+        }
+    }
+}
+
+impl Clone for InMemoryStateImpl {
+    fn clone(&self) -> Self {
+        Self {
+            blocks: RwLock::new(self.blocks.read().clone()),
+            numbers: RwLock::new(self.numbers.read().clone()),
+            pending: RwLock::new(self.pending.read().clone()),
+        }
     }
 }
```

Comment:
~~then we don't need this because we need Clones to be cheap~~

do we even need this to be clone then? if we operate only on `CanonicalInMemoryState`

Reply:
yeah, i added it because of the Clone impl for `BlockchainProvider` https://github.com/paradigmxyz/reth/blob/08c8d251b737e0674ee3d6c62b7d7cef57ad8d9e/crates/storage/provider/src/providers/mod.rs#L87-L88 if we are going to add a `CanonicalInMemoryState` field we need to be able to clone it, and cloning `CanonicalInMemoryState` requires cloning `InMemoryStateImpl`?

Reply:
hmm, but CanonicalInMemoryState is just an arc wrapper so this should be clone?

https://github.com/paradigmxyz/reth/pull/9588/files#diff-015bad1cc2f99d1859913f10cd24ae539d4d5bc4eab4359ff196991c53f95a22R223-R225

Reply:
you are totally right, thx! fixed here https://github.com/paradigmxyz/reth/pull/9588/commits/70521ff870e0a2e71fea537d8db70d9f2b3bc9fa